### ISSUE TITLE
MV2.filled does not preserve type

### ISSUE THREAD

**Issue by lee1043** [CDAT/cdms#270]
```python
import MV2
x = MV2.array([1,2,3])
type(x)
```
Type of x is `<class 'cdms2.tvariable.TransientVariable'>`
```python
y = MV2.filled(x, fill_value=1)
type(y)
```
Type of y is `<type 'numpy.ndarray'>`. 

In my understanding `MV2.filled` is supposed to return input array itself in this case because the input is NOT a masked array (see [documentation](https://cdms.readthedocs.io/en/docstanya/generated/MV2.filled.html)). However, after apply `MV2.filled`, type as MV2 array is changed to regular numpy array (even when given input is masked array). Not sure if it is meant to be or bug. What I prefer to have is to keep the variable type as same before and after. If it is meant to be is there any alternative way? 

Asking for help from @dnadeau4 or @doutriaux1. Thanks!

**Comment by lee1043** [comment]
@dnadeau4 says it should be returned as `TransientVariable` so I am marking this issue as bug.

**Comment by jypeter** [comment]
Hmmm, I'm not sure this should be considered as a bug! I have used for years the fact that is **x** is an MV2 variable, I can easily and reliably convert it to a regular (and non masked) numpy matrix with `x.filled()`, which I have to use when I pass an MV2 variable to **matplotlib** (otherwise I get weird exceptions).
- `x.filled()` **should** return a numpy array
- I'm not against the fact that `MV2.filled(x)` _could_ return an MV2 object, because your are explicitly calling MV2

Some code below to show the problems one can get when moving between masked arrays or variables and arrays (masked or not). That's with CDAT 2.12 and maybe it's not working the same in CDAT 8.

Sorry it's a bit long, but it's something that has bothered me for a long time, and I have tried to pinpoint it in an example at last. Maybe there should be a tutorial about that
```
>>> import numpy as np, cdms2, MV2
>>> id3 = MV2.identity(3)
# Apparently at this stage, we still have a np.ma object, even if it was created with MV2
>>> type(id3)
<class 'numpy.ma.core.MaskedArray'>
>>> type(id3.filled())
<type 'numpy.ndarray'>
# id3 has no masked element, so the operations below with 'np' are 'safe'
>>> type(np.array(id3))
<type 'numpy.ndarray'>
>>> type(np.asarray(id3))
<type 'numpy.ndarray'>
# It is of course better and safer to use 'np.ma'
>>> type(np.ma.array(id3))
<class 'numpy.ma.core.MaskedArray'>
>>> type(np.ma.asarray(id3))
<class 'numpy.ma.core.MaskedArray'>

# Now we actually mask an element of the array, but we still have a np.ma object
>>> id3[1, 1] = MV2.masked
>>> id3
masked_array(data =
 [[1.0 0.0 0.0]
 [0.0 -- 0.0]
 [0.0 0.0 1.0]],
             mask =
 [[False False False]
 [False  True False]
 [False False False]],
       fill_value = 1e+20)
>>> type(id3)
<class 'numpy.ma.core.MaskedArray'>
# Using np.array on an array that has masked elements is of course the wrong thing to do
# but we can't really help it if the user does not now about masks (there should probably
# be a tutorial about this, since the masks are an asset of cdms2!)
>>> type(np.array(id3))
<type 'numpy.ndarray'>
>>> type(np.asarray(id3))
<type 'numpy.ndarray'>
>>> np.array(id3)
array([[ 1.,  0.,  0.],
       [ 0.,  1.,  0.],
       [ 0.,  0.,  1.]])
# This is probably what the user want (rather than what we get above)
>>> type(id3.filled())
<type 'numpy.ndarray'>
>>> id3.filled()
array([[  1.00000000e+00,   0.00000000e+00,   0.00000000e+00],
       [  0.00000000e+00,   1.00000000e+20,   0.00000000e+00],
       [  0.00000000e+00,   0.00000000e+00,   1.00000000e+00]])
# Things are of course correct when we use np.ma, and we still get a np.ma array
# when we use np.ma.array, or np.ma.asarray
>>> type(np.ma.array(id3))
<class 'numpy.ma.core.MaskedArray'>
>>> type(np.ma.asarray(id3))
<class 'numpy.ma.core.MaskedArray'>
>>> np.ma.array(id3)
masked_array(data =
 [[1.0 0.0 0.0]
 [0.0 -- 0.0]
 [0.0 0.0 1.0]],
             mask =
 [[False False False]
 [False  True False]
 [False False False]],
       fill_value = 1e+20)

# Now we create at last a TransientVariable
>>> vid3 = MV2.masked_array(id3)
>>> type(vid3)
<class 'cdms2.tvariable.TransientVariable'>
>>> type(vid3.filled())
<type 'numpy.ndarray'>
# Of course we still get something wrong if we use 'np' and not 'np.ma' or '.filled()'
# BUT AT LEAST WE GET A NUMPY ARRAY
>>> type(np.array(vid3))
<type 'numpy.ndarray'>
>>> type(np.asarray(vid3))
<type 'numpy.ndarray'>
>>> np.array(vid3)
array([[ 1.,  0.,  0.],
       [ 0.,  1.,  0.],
       [ 0.,  0.,  1.]])
# Now, if we want to only get the np.ma part of the MV2 variable, we have 2 ways that work
# and ONE WAY THAT DOES NOT WORK. The problem is that I never remember which one
# and that's why I use x.filled() to be sure to get a np variable (even if I lose the mask)
# Besides, I'm not sure lots of users know about the 'raw=1' trick that I tend to forget myself
>>> type(np.ma.array(vid3))
<class 'cdms2.tvariable.TransientVariable'>
>>> type(np.ma.asarray(vid3))
<class 'numpy.ma.core.MaskedArray'>
>>> type(vid3(raw=1))
<class 'numpy.ma.core.MaskedArray'>
```

**Comment by doutriaux1** [comment]
I agree I'm pretty sure LOTS of code relies on `variable.filled()` to come back as a numpy array.
also maksed arrays come back as numpy when `filled()`

```python
In [1]: import numpy.ma
In [2]: a = numpy.ma.array([1,2,3,4,numpy.ma.masked])
/Users/doutriaux1/anaconda2/envs/cdat/lib/python2.7/site-packages/numpy/ma/core.py:2809: UserWarning: Warning: converting a masked element to nan.
  order=order, subok=True, ndmin=ndmin)

In [3]: type(a.filled())
Out[3]: numpy.ndarray

**Comment by lee1043** [comment]
@jypeter @doutriaux1 thanks for following up. If it is meant to be, I have no problem for leaving it as is. I just was not clear, good to learn.

**Comment by doutriaux1** [comment]
@lee1043 I'm still not sure about `MV2.filled(variable)` though. I think it's nice as is because it's consistent with `variable.filled()`, but it is also true that this could/should be equivalent to `MV2.array(variable.filled())` . What to do with the axes though? Should it be instead the equivalent of 
`MV2.where(variable.masked, fill_value, variable)` ?

**Comment by doutriaux1** [comment]
I'll let @dnadeau4 decide. Re-opening until then.

**Comment by lee1043** [comment]
To keep both array and axes what I was doing is looking like below:
```python
import MV2
d = MV2.array([0,1,2,3,4])
type(d)
```
`cdms2.tvariable.TransientVariable`
```python
d2 = MV2.filled(d, fill_value=1.)
type(d2)
```
`numpy.ndarray`
```python
d2 = MV2.array(d2)
type(d2)
```
`cdms2.tvariable.TransientVariable`
```python
d2.setAxis(0,d.getAxis(0))
```

**Comment by durack1** [comment]
I'll chime in, I think that preserving the type (and attributes) of whatever is sent is key, so this means that you can pass a transientVariable which preserves the axis info etc, having to reattach these after the fact is unnecessary in my opinion. It's a trivial single line command to convert from a transientVariable to numpy array if you want, it's many more lines to reinstate the transientVariable with all attributes intact, so preservation is key

**Comment by doutriaux1** [comment]
@durack1 @lee1043 so it sounds like 
```python
MV2.filled(variable)
# should be equivalent to
MV2.where(variable.masked, fill_value, variable)
```

**Comment by lee1043** [comment]
If many codes were already rely on current way of `MV2.filled`, changing it may cause some users get frustrated when cdms update break their scripts. It is also true what @durack1 says, which I was initially arguing for. One thing for sure is, we will need to clarify this, either way we decide, in the cdms document.

**Comment by doutriaux1** [comment]
@lee1043 no I think codes rely on `variable.filled()` not `MV2.filled(variable)`

**Comment by durack1** [comment]
My suggestion, if it makes more sense and is improves what a user would expect for behaviours then we change and document this. There are many things that could be updated to provide a more logical analysis pathway and simplifying and making the behaviours more consistent would benefit existing and new users.

**Comment by doutriaux1** [comment]
@durack1 yes, so I would expect `MV2.filled` to return an `MV2` and `variable.filled()` to return a `numpy` to be consistent with what `numpy.ma` does.

**Comment by lee1043** [comment]
@doutriaux1 like the idea. My vote for it.

**Comment by dnadeau4** [comment]
I agree with @doutriaux1 and @durack1 and that is what I will implement.  MV2 with returned an MV2, but if variable is of type MV2 then calling variable.filled() than it will return a MV2. 

@doutriaux1 what is the type of variable: `type(variable)`

**Comment by doutriaux1** [comment]
@dnadeau4 I'm not sure which `variable` we're talking about at this point, such a long thread.

**Comment by doutriaux1** [comment]
@dnadeau4 variable.filled() shoud return a pure numpy array. Because that's what masked array do if you call `.filled()` on a masked numpy array it returns a pure numpy array.

**Comment by dnadeau4** [comment]
@doutriaux1 if variable is a "transient variable",   MV2.filled() will return a transient variable with the mask=None.  You will not loose the grid an other attributes this way.

**Comment by jypeter** [comment]
This has become quite a long thread indeed! @doutriaux1 and @dnadeau4 can you look at least at the very end of [my discussion above ](https://github.com/CDAT/cdms/issues/270#issuecomment-416887822)about the problems when moving between np, np.ma and MV2, and check if the following is a bug or a feature? Thanks!
```
>>> type(np.ma.array(vid3))
<class 'cdms2.tvariable.TransientVariable'>
>>> type(np.ma.asarray(vid3))
<class 'numpy.ma.core.MaskedArray'>
>>> type(vid3(raw=1))
<class 'numpy.ma.core.MaskedArray'>
```

**Comment by dnadeau4** [comment]
@jypeter ok I looked more carefully at your example, and it makes sense that `MV2.filled()` returns a `ndarray` since the mask is gone and filled.

@doutriaux1 
is `MV2.where(variable.masked, fill_value, variable)`  equivalent to numpy.ma.where(MV2.where(variable.masked, fill_value, variable)
??

**Comment by doutriaux1** [comment]
@dnadeau4 yes MV2 funcs are in theory equivalent to their numpy.ma where possible.
